org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServers.java

```diff
@@ -268,6 +269,42 @@ public long getStartVersion() {
 			return this.startVersion;
 		}
 
+		public SingleServerDocumentExecutor toExecutor(final @NonNull LanguageServerWrapper serverWrapper) {
+			return new SingleServerDocumentExecutor(this.document, serverWrapper);
+		}
+	}
+
+	/**
+	 * Executor for a single server - for follow-up requests that should go to the server that returned an earlier response
+	 *
+	 */
+	public static class SingleServerDocumentExecutor extends LanguageServerDocumentExecutor {
+
+		private final @NonNull LanguageServerWrapper serverWrapper;
+
+		SingleServerDocumentExecutor(final @NonNull IDocument document, final @NonNull LanguageServerWrapper serverWrapper) {
+			super(document);
+			this.serverWrapper = serverWrapper;
+		}
+
+		@Override
+		@SuppressWarnings("null")
+		protected @NonNull List<@NonNull CompletableFuture<@Nullable LanguageServerWrapper>> getServers() {
+			// Compute list of servers from document & filter
+			return Collections.singletonList(CompletableFuture.completedFuture(this.serverWrapper));
+		}
+
+		public <T> CompletableFuture<T> execute(@NonNull Function<LanguageServer, ? extends CompletionStage<T>> fn) {
+			return this.serverWrapper.execute(fn);
+		}
+
+		public boolean supports(final Predicate<ServerCapabilities> predicate) {
```

Comment:
I think it not right that this executor does not respect the `withFilter` or `withPredicate` (even though one could call it) and instead offers a new method to do the filter

Reply:
See summary of revisions below